// @hprose/rpc-node/lib/UdpTransport.js
"use strict";
/*--------------------------------------------------------*\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: https://hprose.com                     |
|                                                          |
| UdpTransport.ts                                          |
|                                                          |
| UdpTransport for TypeScript.                             |
|                                                          |
| LastModified: Dec 18, 2019                               |
| Author: Ma Bingyao <andot@hprose.com>                    |
|                                                          |
\*________________________________________________________*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UdpTransport = void 0;
var dgram = __importStar(require("dgram"));
var url_1 = require("url");
var rpc_core_1 = require("@hprose/rpc-core");
var io_1 = require("@hprose/io");
var UdpTransport = /** @class */ (function () {
    function UdpTransport() {
        this.counter = 0;
        this.results = new Map();
        this.sockets = Object.create(null);
    }
    UdpTransport.prototype.getSocket = function (uri) {
        return __awaiter(this, void 0, void 0, function () {
            var socket, udp, parser, protocol, type, onerror;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.sockets[uri]];
                    case 1:
                        socket = _a.sent();
                        if (socket !== undefined) {
                            return [2 /*return*/, socket];
                        }
                        udp = rpc_core_1.defer();
                        parser = url_1.parse(uri);
                        protocol = parser.protocol;
                        type = 'udp4';
                        switch (protocol) {
                            case 'udp:':
                            case 'udp4:':
                                break;
                            case 'udp6:':
                                type = 'udp6';
                                break;
                            default:
                                throw new Error('unsupported ' + protocol + ' protocol');
                        }
                        socket = dgram.createSocket(type);
                        socket.unref();
                        socket.on('listening', function () {
                            udp.resolve(socket);
                        });
                        socket.on('message', function (msg) { return __awaiter(_this, void 0, void 0, function () {
                            var crc, header, bodyLength, index, has_error, response, results, result;
                            return __generator(this, function (_a) {
                                crc = msg.readInt32BE(0);
                                header = msg.subarray(4, 8);
                                if (rpc_core_1.crc32(header) !== crc)
                                    return [2 /*return*/];
                                bodyLength = msg.readUInt16BE(4);
                                if (bodyLength !== msg.length - 8)
                                    return [2 /*return*/];
                                index = msg.readUInt16BE(6);
                                has_error = (index & 0x8000) !== 0;
                                index &= 0x7FFF;
                                response = new Uint8Array(msg.buffer, msg.byteOffset + 8, bodyLength);
                                results = this.results.get(socket);
                                if (results) {
                                    result = results[index];
                                    delete results[index];
                                    if (result) {
                                        if (has_error) {
                                            result.reject(new Error(io_1.fromUint8Array(response)));
                                        }
                                        else {
                                            result.resolve(response);
                                        }
                                    }
                                }
                                return [2 /*return*/];
                            });
                        }); });
                        onerror = function (error) { return __awaiter(_this, void 0, void 0, function () {
                            var results, index, result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        results = this.results.get(socket);
                                        if (results) {
                                            for (index in results) {
                                                result = results[index];
                                                result.reject(error);
                                                delete results[index];
                                            }
                                        }
                                        return [4 /*yield*/, this.sockets[uri]];
                                    case 1:
                                        if ((_a.sent()) === socket) {
                                            delete this.sockets[uri];
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        }); };
                        socket.on('error', onerror);
                        socket.on('close', function () { return onerror(new Error('closed')); });
                        socket.bind();
                        this.sockets[uri] = udp.promise;
                        return [2 /*return*/, udp.promise];
                }
            });
        });
    };
    UdpTransport.prototype.transport = function (request, context) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var uri, index, result, socket, results, timeoutId_1, parser, n, header, crc, body;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (request.length > 65499) {
                            throw new Error('request too large');
                        }
                        uri = context.uri;
                        index = (this.counter < 0x7FFF) ? ++this.counter : this.counter = 0;
                        result = rpc_core_1.defer();
                        return [4 /*yield*/, this.getSocket(uri)];
                    case 1:
                        socket = _b.sent();
                        if (this.results.get(socket) === undefined) {
                            this.results.set(socket, Object.create(null));
                        }
                        results = this.results.get(socket);
                        results[index] = result;
                        if (context.timeout > 0) {
                            timeoutId_1 = setTimeout(function () {
                                delete results[index];
                                result.reject(new rpc_core_1.TimeoutError());
                            }, context.timeout);
                            result.promise.then(function () {
                                clearTimeout(timeoutId_1);
                            }, function () {
                                clearTimeout(timeoutId_1);
                            });
                        }
                        parser = url_1.parse(uri);
                        n = request.length;
                        header = Buffer.allocUnsafe(8);
                        header.writeUInt16BE(n, 4);
                        header.writeUInt16BE(index, 6);
                        crc = rpc_core_1.crc32(new Uint8Array(header.buffer, header.byteOffset + 4, 4));
                        header.writeInt32BE(crc, 0);
                        body = Buffer.from(request.buffer, request.byteOffset, request.length);
                        socket.send([header, body], parser.port ? parseInt(parser.port, 10) : 8412, (_a = parser.hostname) !== null && _a !== void 0 ? _a : undefined, function (error) {
                            if (error) {
                                delete results[index];
                                result.reject(error);
                            }
                        });
                        return [2 /*return*/, result.promise];
                }
            });
        });
    };
    UdpTransport.prototype.abort = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _i, uri, socket;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = [];
                        for (_b in this.sockets)
                            _a.push(_b);
                        _i = 0;
                        _c.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 4];
                        uri = _a[_i];
                        socket = this.sockets[uri];
                        delete this.sockets[uri];
                        if (!socket) return [3 /*break*/, 3];
                        return [4 /*yield*/, socket];
                    case 2:
                        (_c.sent()).close();
                        _c.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    UdpTransport.schemes = ['udp', 'udp4', 'udp6'];
    return UdpTransport;
}());
exports.UdpTransport = UdpTransport;
rpc_core_1.Client.register('udp', UdpTransport);
