// @hprose/io/lib/index.js
"use strict";
/*--------------------------------------------------------*\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: https://hprose.com                     |
|                                                          |
| hprose.io.ts                                             |
|                                                          |
| @hprose/io for TypeScript.                               |
|                                                          |
| LastModified: Jan 22, 2019                               |
| Author: Ma Bingyao <andot@hprose.com>                    |
|                                                          |
\*________________________________________________________*/
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ByteStream_1 = require("./ByteStream");
exports.ByteStream = ByteStream_1.ByteStream;
exports.writeInt32BE = ByteStream_1.writeInt32BE;
exports.writeInt32LE = ByteStream_1.writeInt32LE;
exports.fromUint8Array = ByteStream_1.fromUint8Array;
exports.toBinaryString = ByteStream_1.toBinaryString;
var Writer_1 = require("./Writer");
exports.Writer = Writer_1.Writer;
var Reader_1 = require("./Reader");
exports.Reader = Reader_1.Reader;
var TypeManager = __importStar(require("./TypeManager"));
exports.TypeManager = TypeManager;
var Serializer = __importStar(require("./Serializer"));
exports.Serializer = Serializer;
var Deserializer = __importStar(require("./Deserializer"));
exports.Deserializer = Deserializer;
var Formatter = __importStar(require("./Formatter"));
exports.Formatter = Formatter;
var ReferenceReader = __importStar(require("./ReferenceReader"));
exports.ReferenceReader = ReferenceReader;
var ValueReader = __importStar(require("./ValueReader"));
exports.ValueReader = ValueReader;
var ValueWriter = __importStar(require("./ValueWriter"));
exports.ValueWriter = ValueWriter;
