// @hprose/rpc-core/lib/PluginManager.js
"use strict";
/*--------------------------------------------------------*\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: https://hprose.com                     |
|                                                          |
| PluginManager.ts                                         |
|                                                          |
| PluginManager for TypeScript.                            |
|                                                          |
| LastModified: Feb 16, 2020                               |
| Author: Ma Bingyao <andot@hprose.com>                    |
|                                                          |
\*________________________________________________________*/
Object.defineProperty(exports, "__esModule", { value: true });
var PluginManager = /** @class */ (function () {
    function PluginManager(defaultHandler) {
        this.defaultHandler = defaultHandler;
        this.handlers = [];
        this.firstHandler = defaultHandler;
    }
    PluginManager.prototype.rebuildHandler = function () {
        var handlers = this.handlers;
        var next = this.defaultHandler;
        var n = handlers.length;
        for (var i = n - 1; i >= 0; --i) {
            next = this.getNextHandler(handlers[i], next);
        }
        this.firstHandler = next;
    };
    Object.defineProperty(PluginManager.prototype, "handler", {
        get: function () {
            return this.firstHandler;
        },
        enumerable: true,
        configurable: true
    });
    PluginManager.prototype.use = function () {
        var _a;
        var handlers = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            handlers[_i] = arguments[_i];
        }
        (_a = this.handlers).push.apply(_a, handlers);
        this.rebuildHandler();
    };
    PluginManager.prototype.unuse = function () {
        var handlers = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            handlers[_i] = arguments[_i];
        }
        var rebuild = false;
        for (var i = 0, n = handlers.length; i < n; ++i) {
            var index = this.handlers.indexOf(handlers[i]);
            if (index >= 0) {
                this.handlers.splice(index, 1);
                rebuild = true;
            }
        }
        if (rebuild)
            this.rebuildHandler();
    };
    return PluginManager;
}());
exports.PluginManager = PluginManager;
